### PR TITLE
virt: Detect proot virtualisation by ptrace metadata

### PR DESCRIPTION
**Pull request by cdown**
proot provides userspace-powered emulation of chroot and mount --bind,
lending it to be used on environments without unprivileged user
namespaces, or in otherwise restricted environments like Android.

In order to achieve this, proot makes use of the kernel's ptrace()
facility, which we can use in order to detect its presence. Since it
doesn't use any kind of namespacing, including PID namespacing, we don't
need to do any tricks when trying to get the tracer's metadata.

For our purposes, proot is listed as a "container", since we mostly use
this also as the bucket for non-container-but-container-like
technologies like WSL. As such, it seems like a good fit for this
section as well.

Also see https://github.com/proot-me/proot/issues/217.

### CODE REVIEW

**Comment by cdown** [comment]
(for some reason travis is picking up the wrong commit, force pushed again.)

**Comment by cdown** [comment]
- amd64 timed out trying to run upstream tests.
- i386 logind tests failed for some reason, but seems really unlikely it's related.

File: src/basic/virt.c [hunk 1]
```diff
@@ -472,6 +474,22 @@ int detect_container(void) {
...
+        /* proot doesn't use PID namespacing, so we can just check if we have a matching tracer for this
+         * invocation without worrying about it being elsewhere.
+         */
+        r = get_proc_field("/proc/self/status", "TracerPid", WHITESPACE, &p);
+        if (r == 0 && !streq(p, "0")) {
+                pid_t ptrace_pid;
+                r = parse_pid(p, &ptrace_pid);
+                if (r == 0) {
+                        const char *pf = procfs_file_alloca(ptrace_pid, "comm");
+                        _cleanup_free_ char *ptrace_comm;
```
Comment (keszybz) [inline src/basic/virt.c]: This needs to be initialized to avoid freeing an unrelated pointer if read_one_line_file() fails below.

**Comment by cdown** [comment]
Thanks for the review!

**Pull request merged**